libsolidity/codegen/ContractCompiler.cpp

```diff
@@ -441,7 +440,7 @@ void ContractCompiler::appendFunctionSelector(ContractDefinition const& _contrac
 		}
 		else
 			// TODO: error message here?
-			m_context.appendRevert();
+			m_context.appendRevert("Contract cannot receive Ether");
```

Comment:
This is the other one? I don't think this is actually ``Contract cannot receive Ether``, but rather "Unknown signature and no fallback defined." - and only if you *do* have a receive function.
So I reckon it happens for one of or both of these (the first might be caught elsewhere in the dispatch, but at least the second with non-empty but short calldata works - that's the jump in L390):
```
contract C {
  receive() external {}
}
// ====
// (): 0x42 -> FAILURE
// (): hex"0x42" -> FAILURE
```

You do realize that an easy way to find out how to trigger lines like that is to just comment them out and run the test suite? Here this fails for the second to last call in ``semanticTests/receive/empty_calldata_calls_receive.sol``, so that triggers it... and if there's no test failing, then there's something fishy anyways...

Reply:
But look, if there is no fallback and no receive, it shouldn't go into the outer if at all.
If there is one of those, they add a `STOP`. So how does it reach that `REVERT`?!

Reply:
Yes, the C++ line is triggered, but I meant the `REVERT` instruction it adds is unreachable because of the previous `STOP`s. That's why no test fails I would say.

Reply:
Look at the code with added comments about the possible code paths:
```
m_context << u256(4) << Instruction::CALLDATASIZE << Instruction::LT; // This can be true for 0, 1, 2 or 3 bytes of calldata, in which case we jump to "notFoundOrReceiveEther"
m_context.appendConditionalJumpTo(notFoundOrReceiveEther);

// so until we reach "notFoundOrReceiveEther" we're in the case "calldatasize >= 4"

// If there is no interface functions, we skip right over this whole thing.
// If there is some, we add entry points, but we still might fall out of this
// ``if``, if there is no match I think (haven't *really* checked that, though).
if (!interfaceFunctions.empty())
{
	CompilerUtils(m_context).loadFromMemory(0, IntegerType(CompilerUtils::dataStartOffset * 8), true);

	// stack now is: <can-call-non-view-functions>? <funhash>
	vector<FixedHash<4>> sortedIDs;
	for (auto const& it: interfaceFunctions)
	{
		callDataUnpackerEntryPoints.emplace(it.first, m_context.newTag());
		sortedIDs.emplace_back(it.first);
	}
	std::sort(sortedIDs.begin(), sortedIDs.end());
	appendInternalSelector(callDataUnpackerEntryPoints, sortedIDs, notFound, m_optimiserSettings.expectedExecutionsPerDeployment);
}

// So here we get out in one of two cases:
// - calldatasize >= 4 and no matching interface function (at least probably)
// - and definitely for calldatasize < 4 
m_context << notFoundOrReceiveEther;

// Now there is only *one* case in which the revert in question is hit,
// namely assume there *is* a receive function, but *no* fallback function.

if (!fallback && !etherReceiver) // This is "false" in that case, so onto the ``else``
	m_context.appendRevert();
else
{
	if (etherReceiver) // This is "true" in our case.
	{
		// This is the critical jump now: if we have ``calldatasize > 0``. But no reason that we shouldn't.
		m_context << Instruction::CALLDATASIZE;
		m_context.appendConditionalJumpTo(notFound); // So in that case we jump to ``notFound``

               // so we can ignore all this up to ``notFound`` (and jumping *over* the stop you mentioned)

		solAssert(!_contract.isLibrary(), "");
		solAssert(etherReceiver->isReceive(), "");
		solAssert(FunctionType(*etherReceiver).parameterTypes().empty(), "");
		solAssert(FunctionType(*etherReceiver).returnParameterTypes().empty(), "");
		etherReceiver->accept(*this);
		m_context << Instruction::STOP;
	}

        // So we end up here.
	m_context << notFound;
	if (fallback) // But we *don't* have a fallback
	{
		solAssert(!_contract.isLibrary(), "");
		if (!fallback->isPayable() && needToAddCallvalueCheck)
			appendCallValueCheck();

		solAssert(fallback->isFallback(), "");
		solAssert(FunctionType(*fallback).parameterTypes().empty(), "");
		solAssert(FunctionType(*fallback).returnParameterTypes().empty(), "");
		fallback->accept(*this);
		m_context << Instruction::STOP;
	}
	else
		// So finally we *do* arrive here...
		m_context.appendRevert();
}
```

Reply:
Satisfied by that ;-)?

Reply:
In short: You jump over the ``STOP`` added in the receive case, in case calldata is non-empty.

Reply:
Ah true, missed the jump. Ok, will try to trigger the failure somehow

Reply:
As I said - this should definitely do:
```
contract C {
  receive() external {}
}
// ====
// (): hex"0x42" -> FAILURE, ...
```